[WIP] Add Topic Configuration

cricket007 opened the pull request:
- [x] Parse YAML to `NewTopic` Kafka object
- [ ] ~Make topic factory discoverable(?)~
- [x] Pass variables between AdminClientFactory and Topic factory to create topics
- [ ] Integration Test to create topics

mzamani1 commented on src/main/java/io/dropwizard/kafka/KafkaTopicsFactory.java:
```diff
@@ -0,0 +1,61 @@
+        private short replicationFactor;
+
+        @JsonProperty
+        Map<String, String> configs = null;
```
Instead of `null`, maybe default it to `Collections.emptyMap()`?

cricket007 replied:
It's null in the Kafka source code. I think it should be okay.

mzamani1 replied:
That comment was more in the interest of following the conventions I usually follow for these libraries. 

I tend to prefer default null objects over optionals (with null checks in the code), but empty collections by default. 

Not super strict on that, and I'm willing to let it slide, but it will differ from the other stuff in this library.

mzamani1 commented on src/main/java/io/dropwizard/kafka/KafkaTopicsFactory.java:
```diff
@@ -0,0 +1,61 @@
+
+public class KafkaTopicsFactory extends ArrayList<KafkaTopicsFactory.KafkaTopic> implements Discoverable {
+
+    static class KafkaTopic {
```
Would this be cleaner to split this out to a separate class?

mzamani1 commented on src/main/java/io/dropwizard/kafka/KafkaTopicsFactory.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class KafkaTopicsFactory extends ArrayList<KafkaTopicsFactory.KafkaTopic> implements Discoverable {
```
I'd say instead of inheritance, we should prefer composition.

The `KafkaTopicsFactory` can contain a field that is `List<KafkaTopic>`

cricket007 replied:
I agree extending looks bad, though I was aiming for the Configuration file not needing to look like this

```java
public class ExampleConfiguration {
    KafkaTopicsFactory topics;
}
```

```yaml
topics:
   topics: 
     - name: ...
```

Is there  a way around that nesting?

mzamani1 replied:
Not the way it's currently structured.

Why not restructure it such that it looks more like this:

```java
public class ExampleConfiguration {
    @Valid
    @MinSize(1)
    @JsonProperty
    List<KafkaTopicsFactory> topics;
}
```

```yaml
topics:
  - name: ...
```

mzamani1 replied:
I.e., instead of having the `KafkaTopicsFactory` contain the list, just have a _list of_ `KafkaTopicsFactory`s in your app's configuration.

And, if you go that route, maybe rename it `KafkaTopicFactory` (remove the plural for topic)

mzamani1 commented on src/main/java/io/dropwizard/kafka/KafkaTopicsFactory.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class KafkaTopicsFactory extends ArrayList<KafkaTopicsFactory.KafkaTopic> implements Discoverable {
```
This probably also doesn't need to be `Discoverable`. That's usually only useful when you have multiple implementations of a factory that you want to be able to switch to based on some value in the YAML, and not (ideally) having to change the code.

mzamani1 commented on src/main/java/io/dropwizard/kafka/lifecycle/TopicCreationLifecycleListener.java:
```diff
@@ -0,0 +1,23 @@
+
+import java.util.Collection;
+
+public abstract class TopicCreationLifecycleListener implements LifeCycle.Listener {
```
Instead of Jetty's lifecycle, you should use Dropwizard's abstraction on top of it, `Managed`.

cricket007 replied:
Okay, thought I needed `environment.lifecycle().addListener()`

mzamani1 replied:
For these type of modules, I usually prefer `environment.lifecycle().manage(...)` instead, unless there's a reason for it not to be done that way.

mzamani1 commented on src/main/resources/META-INF/services/io.dropwizard.jackson.Discoverable:
```diff
@@ -1,3 +1,4 @@
+io.dropwizard.kafka.KafkaTopicsFactory
```
Probably not needed

mzamani1 commented:
@cricket007 any update on this?

cricket007 commented:
Sorry, was on vacation. Addressed the review feedback. Not sure how to get the YAMLConfigurationFactory to test a list of topics, but I can do without that. 

Regarding the actual creation of topics, I can't seem to get `DropwizardKafkaIT` to run with `mvn integration-test`

mzamani1 commented on src/test/java/io/dropwizard/kafka/integration/AdminClientIntegrationConfiguration.java:
```diff
@@ -0,0 +1,34 @@
+
+    @NotNull
+    @JsonProperty("topic")
+    private KafkaTopicFactory topic;
```
2 comments:

1. I think the `KafkaTopicFactory` should maybe be contained **in the** `BasicKafkaAdminClientFactory` as a member. 
2. Either the `KafkaTopicFactory` should be able to be `null`, or it should have a boolean flag `enabled` inside of it to allow opting out of that functionality.

mzamani1 commented on pom.xml:
```diff
@@ -135,6 +135,22 @@
+    <plugins>
+      <plugin>
+        <artifactId>maven-failsafe-plugin</artifactId>
+        <version>3.0.0-M3</version>
```
I'm going to commit this in a separate PR, so I'll just have you rebase against that if you don't mind.

cricket007 replied:
Sure. The tests don't work on my Mac anyway. Says something is wrong with the `log.dir` property.

Do you get the same? I wasn't able to quickly find any similar errors searching the spring kafka issues

mzamani1 replied:
@cricket007 When building from master, I don't get those issues. Have you rebased this commit against the master changes?

mzamani1 replied:
I cloned your branch locally, and I do get this error when running a build:

```
14:04:48.441 [main] ERROR kafka.server.BrokerMetadataCheckpoint - Failed to read meta.properties file under dir /var/folders/b7/j31znfln6gvc4sghhcq4nwqr0000gn/T/kafka-1473154000058593224/meta.properties due to /var/folders/b7/j31znfln6gvc4sghhcq4nwqr0000gn/T/kafka-1473154000058593224/meta.properties
14:04:48.443 [main] ERROR kafka.server.KafkaServer - Fail to read meta.properties under log directory /var/folders/b7/j31znfln6gvc4sghhcq4nwqr0000gn/T/kafka-1473154000058593224
java.nio.file.NoSuchFileException: /var/folders/b7/j31znfln6gvc4sghhcq4nwqr0000gn/T/kafka-1473154000058593224/meta.properties
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
	at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
	at java.nio.file.Files.newByteChannel(Files.java:361)
	at java.nio.file.Files.newByteChannel(Files.java:407)
	at java.nio.file.spi.FileSystemProvider.newInputStream(FileSystemProvider.java:384)
	at java.nio.file.Files.newInputStream(Files.java:152)
	at org.apache.kafka.common.utils.Utils.loadProps(Utils.java:560)
	at kafka.server.BrokerMetadataCheckpoint.liftedTree2$1(BrokerMetadataCheckpoint.scala:63)
	at kafka.server.BrokerMetadataCheckpoint.read(BrokerMetadataCheckpoint.scala:62)
	at kafka.server.KafkaServer$$anonfun$getBrokerIdAndOfflineDirs$1.apply(KafkaServer.scala:668)
	at kafka.server.KafkaServer$$anonfun$getBrokerIdAndOfflineDirs$1.apply(KafkaServer.scala:666)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:35)
	at kafka.server.KafkaServer.getBrokerIdAndOfflineDirs(KafkaServer.scala:666)
	at kafka.server.KafkaServer.startup(KafkaServer.scala:209)
	at kafka.utils.TestUtils$.createServer(TestUtils.scala:132)
	at kafka.utils.TestUtils.createServer(TestUtils.scala)
	at org.springframework.kafka.test.EmbeddedKafkaBroker.afterPropertiesSet(EmbeddedKafkaBroker.java:223)
	at org.springframework.kafka.test.rule.EmbeddedKafkaRule.before(EmbeddedKafkaRule.java:109)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:46)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
```

Perhaps something added when configuring the admin client as well as just producers/consumers is misconfigured in the `EmbeddedKafka`?

cricket007 commented:
Regarding the stacktrace you showed, that's actually a different error than I have, but searching that suggests that since the spring-kafka Kafka dependency version is lower than the other that is actually included. To resolve that, I've moved kafka declaration into dependencyManagement to enforce the newer version, and now I seem to be able to get past that. 

`mvn -Prelease integration-test` now runs as expected, though, not sure how to `startup()` the Managed AdminClient object from the tests such that the topics actually get created.

mzamani1 commented on src/main/java/io/dropwizard/kafka/KafkaAdminClientFactory.java:
```diff
@@ -79,6 +82,11 @@
+    protected boolean createTopics;
+    @Valid
+    @JsonProperty
+    protected Set<KafkaTopicFactory> topics;
```
Maybe default this to an empty set and add a `@NotNull`, just to avoid needing to do null checks. That way, you'll only need to check that the set is empty or not.

mzamani1 commented on src/main/java/io/dropwizard/kafka/KafkaAdminClientFactory.java:
```diff
@@ -79,6 +82,11 @@
     @JsonProperty
     protected SecurityFactory security;
+    @JsonProperty
+    protected boolean createTopics;
```
Maybe rename this `topicCreationEnabled`? I generally like having the `enabled` word present in these type of configs. Also, I know by default this boolean will be `false`, but maybe making that explicit in code would be a little nicer. `protected boolean createTopics = false;`

mzamani1 commented on src/main/java/io/dropwizard/kafka/KafkaAdminClientFactory.java:
```diff
@@ -224,17 +232,43 @@ public void setSecurity(final SecurityFactory security) {
+    }
+
+    protected void manageAdminClient(final LifecycleEnvironment lifecycle, final AdminClient adminClient,
+                                               Collection<NewTopic> topics) {
```
Whitespace. Also maybe qualify `topics` with `final` to maintain consistency.

mzamani1 commented on src/main/java/io/dropwizard/kafka/KafkaTopicFactory.java:
```diff
@@ -0,0 +1,85 @@
+    private short replicationFactor;
+
+    @JsonProperty
+    private Map<String, String> configs = Collections.emptyMap();
```
Maybe add a `@NotNull` to this

mzamani1 commented on src/main/java/io/dropwizard/kafka/managed/KafkaAdminClientManager.java:
```diff
@@ -12,15 +15,20 @@
     @Override
     public void start() throws Exception {
-        // do nothing
+        if (this.topics != null && !this.topics.isEmpty()) {
```
If `topics` can be non-null, this can be simplified a bit.

mzamani1 commented on src/main/java/io/dropwizard/kafka/managed/KafkaAdminClientManager.java:
```diff
@@ -12,15 +15,20 @@
-        // do nothing
+        if (this.topics != null && !this.topics.isEmpty()) {
+            // TODO: Check if topics already exist
+            this.adminClient.createTopics(this.topics);
```
Out of curiosity, what is the behavior of the client if `createTopics` is created every time the app starts up? Will it be a no-op if the topics already exist? Otherwise, the check might be pretty important.

mzamani1 commented:
@cricket007 hey, let me know if you’re going to circle back to this today or tomorrow. If not I’ll cut a release for some other stuff, and release this when you have time to circle back.

cricket007 commented:
> Should the AdminClientFactory have the Jackson polymorphic deserialization type stuff set up?

Seems to be out of scope of this feature, no? Besides, not sure what what would look like.

In any case, I addressed all the other comments and wrote the simplest thing I could find for "check if topics exist" - https://stackoverflow.com/a/53442463/2308683

mzamani1 commented:
@cricket007 You could argue it’s out of scope for this PR, but it’s absolutely something that we need to support. Either you or I can add that in a separate PR.

mzamani1 commented on src/main/java/io/dropwizard/kafka/managed/KafkaAdminClientManager.java:
```diff
@@ -17,16 +18,27 @@
+                this.topics.removeIf(newTopic -> {
+                    boolean match = newTopic.name().equals(t);
+                    if (match) {
+                        log.warn("Not attempting to re-create existing topic {}.", newTopic.name());
```
I think maybe a debug log (or a single info log with all of the skipper topics appended together) would be more appropriate. Otherwise, every time someone restarts their app, they may be flooded with warn logs.
